### ISSUE TITLE
Pagination change issue

### ISSUE THREAD

**Issue by edgar-pozadas-by** [material-table-core/core#272]
Here you have an example with the error
https://codesandbox.io/s/pagination-zu4gw?file=/src/App.js

When I want to manage the pagination and then pass the data to the table it breaks and show more rows than expected.
It used to work on version 3.0.17 

1. Go to pagination component
2. Change to page 2 and then go back to first one 
3. See error

**Screenshots**
screen shot, on render w
5 rows expected, 5 rows showed 
![image](https://user-images.githubusercontent.com/80122564/126929463-e7a68e18-6077-47d4-aa08-499ef710cb8d.png)
screen shot with the error:
5 rows expected, 8 rows showed
![image](https://user-images.githubusercontent.com/80122564/126929480-6925eab7-3422-4a37-a000-7e843446606d.png)

**Comment by matthewoestreich** [comment]
Thank you for this!

**Comment by Domino987** [comment]
It looks like you are mixing remote and local data. If you have all the data already, you need to pass it as one array. This allows for correct filtering/sorting/pagination. This mixing dos not work. Please use either local or remote data.

**Comment by edgar-arroyo-by** [comment]
Hello there @Domino987 

I have another case in which this is happening:
https://codesandbox.io/s/changin-array-different-ids-jp9bp?file=/src/App.js


Lets say I have two arrays, and i want to programmatically change the array used as data, if the data inside the arrays do not share the same ids, this duplication error happens again.

```
import React from "react";
import MaterialTable from "@material-table/core";

function App() {
  const global_cols = [
    { title: "number", field: "number", minWidth: 140, maxWidth: 400 }
  ];

  const global_data1 = [
    {
      number: "1",
      id: 1
    },
    {
      number: "2",
      id: 2
    }
  ];

  const global_data2 = [
    {
      number: "3",
      id: 7
    },
    {
      number: "4",
      id: 2
    }
  ];

  const [pdata, setPData] = React.useState([]);
  const [switcher, setSwitcher] = React.useState(true);

  React.useEffect(() => {
    if (switcher) setPData(global_data1);
    else setPData(global_data2);
  }, [switcher]);

  return (
    <>
      <button onClick={() => setSwitcher(!switcher)}>Cambiar</button>
      <MaterialTable title="Basic" columns={global_cols} data={pdata} />
    </>
  );
}

export default App;
```

**Comment by edgar-arroyo-by** [comment]
Lets say I have a promise that I resolve on component mount, lets say I save the resulted-array in the state and I used in the table.
Say I have  a button, to recall the promise with different parameters, or to change the called promise. 

It is expected that, if the data inside the resulted-array do not share ids (resulted array that i am resolving before injecting that to the table), it will duplicate? 


thanks in advance


@Domino987  @oze4  @edgar-pozadas-by

**Comment by edgar-arroyo-by** [comment]
Does @material-table/core has opinion on the data ids or in how that arrays are generated?

**Comment by Domino987** [comment]
Hmh, seems weird.
I think this might be caused by `prevDataObject[row.tableData && row.tableData.id]` and shares data between objects that is not supposed to be shared. I will remove this and only listen to row.id, which should remove the problem.
Can you check once we release a new version?

**Comment by edgar-arroyo-by** [comment]
yes sir!
thanks a lot!

**Comment by matthewoestreich** [comment]
@Domino987 I just tested removing that locally, and it does not resolve the error. I imported the CodeSandbox locally to the demos.

From what I have read, it doesn't look like using an index for the key is recommended, and may be why this is happening. For example, if I change the key to the following it works just fine (no error):

```js
key={'row-' + data.tableData.id + '-' + Math.floor(Math.random() * 100)}
```

Perhaps we should create a new prop on the `tableData` object, `tableData.key` or whatever, and actually generate a small unique GUID for each row that we can use internally?

**Comment by edgar-arroyo-by** [comment]
Just adding an (unwanted) opinion:


Removing `prevDataObject[row.tableData && row.tableData.id]` does not solve the issue because the id being used is the index
As @oze4 says, the issue is on using the index as key.  I believe the best way is to generate that small unique GUID and used internally, so the table has no opinion on how the ids of the data should be.

**Comment by matthewoestreich** [comment]
All opinions are wanted! 

I just pushed a commit which contains the added demo for this, if you would like to test locally @Domino987 I called the demo 'DataSwitcher' or something along those lines.

**Comment by Domino987** [comment]
So the problem here is not the index in itself.
It was used for years and your problem might be connected to something else.
Adding Math.random or a random uuid to it will just caused every row to remount on every rerender of the parent component, which will break any state that might be saved for those rows. Thats why the new id token was added to reduce friction.

If you generate a true random id between the tow data sets, the problem should not be there I think but we need to check this with a demo.
Its an interesting problem to solve for sure.

So I just tried out the new demo:
Using a number might overlap the ids causing multiple rows to be render because if no id is give, the id might be index.

```
const tableData = {
        id: row.id || index,
        ...prevTableData,
        ...row.tableData
      };
```
This [code change](https://github.com/material-table-core/core/commit/4dd02633f6639851460feb800272fb953fcb1322) will fallback to index only if no id is given. And the problem should be fixed with this change. Can you check for you as well?

**Comment by edgar-arroyo-by** [comment]
Hello there @Domino987 
Just a question/suggestion

I may be wrong but, wouldn't be a good idea to define a kind of `keyAccessor `prop that by default it is id but can be changed?

Just thinking on what can happen if for example a future user id column is called uid, or something different like objectid 

Sorry, i am just spitballing ideas 

cc @oze4

**Comment by Domino987** [comment]
Hi yes i was actually thinking about this when I implemented this. If this is a thing people want and if it makes the transition easier, yes of course. 

There are several ways to do it. What do you think is the best:
Just a string as accessor key, a function to return the accessor key or a function to return the actual key, or all above ?

**Comment by matthewoestreich** [comment]
@Domino987 In my testing, I was logging renders, and adding the randomness to the end did not increase the amount of rerenders. 

The problem is under the hood React thinks that key already exists, due to the ~~index~~ id being used within the key (which in this case is the same thing as using the index - the data changes, the key doesn't). This is a common problem, [this article](https://robinpokorny.medium.com/index-as-a-key-is-an-anti-pattern-e0349aece318) is directly referenced in the React docs. 

After adding those changes things got worse.  I don't even have to click the button (to load the second set of data) to get the error - now I just get the error right off the bat without doing anything.

**Comment by Domino987** [comment]
Not the amount of rerenders but the amount of mounts/losing of states, sorry for the wrong wording.

It works in my case bit I am open to ideas on how to fix this.

**Comment by edgar-arroyo-by** [comment]
@Domino987  Maybe just something like

`keyAccesor: string | () => string`

**Comment by Domino987** [comment]
> @Domino987  Maybe just something like
> 
> `keyAccesor: string | () => string`

Would you mind creating a PR for this? Looks good to me to get the key. Once this is done, we need to fix the duplicates if that still happens.

**Comment by matthewoestreich** [comment]
In my mind, why not just add a UUID or 'key' prop at that point? Like, lets say we have a function that generates UUID..

```js
const tableData = {
  id: row.id || index,
  uuid: row.uuid || funcToMakeUUID(),
  ...prevTableData,
  ...row.tableData
};
```

**Comment by Domino987** [comment]
I think adding a UUID generated by us would break it for everybody who does not use a id as before the removal of the mutation. Because rerenders would remount all rows every time.

**Comment by matthewoestreich** [comment]
I suppose in that case we could just check if the user supplied an ID or not, if they don't, there should be no difference between the user providing an ID and us generating an ID. I don't see how that would cause any extra renders or mounts.

```js
const tableData = {
  id: row.id || index,
  // Check if user supplied ID, if not, check if we already made one, otherwise make one
  uuid: row.id || row.uuid || funcToMakeUUID(),
  ...prevTableData,
  ...row.tableData
};
```

**Comment by Domino987** [comment]
Actually there would. If they supply the same array, we want the index because the first item from both of those arrays are the same. And we would not find the id in the supplied array because we do not mutate anymore. So we would be generating a new id for the same row and remount it. Previously, the index would be the same and the row would not remount.

**Comment by Domino987** [comment]
But the question is do we want this? Maybe we want to force the user to use the id to remove the problem all together in the future. So an id would be fine and if somebody complains, we say generate an id pls.

**Comment by matthewoestreich** [comment]
I went ahead and made changes and published a [new branch](https://github.com/material-table-core/core/tree/uuid-for-row-key). 

https://github.com/material-table-core/core/blob/807a48413581afad14ed1f06511b626461c05bad/src/utils/data-manager.js#L76-L81

This generates a uuid for each row, I have profiled this and tested but I do not see any extra renders or mounts. This change fixes the issue.

Is there a certain "table config" we can test this rerender theory with? I guess I don't see where this would cause issues since we are only using the uuid for the key - existing comparisons using `.id` will not be changed.

**Comment by oze4** [comment]
Ah ok I think I found what you're talking about! So with that change, when I try to edit a row, that edit doesn't save.

Is that the kind of thing you were talking about @Domino987 ?

**Comment by matthewoestreich** [comment]
Here is how I am reasoning about this issue.

The TLDR is that even if a user provided a unique id from outside the table, and let's say we don't use the index as the key but the user provided id, we can still run into this issue. So yes, technically it's not due to the index as key, but it's still the same exact problem.  The same goes for React. They track each element via the key prop. Thus the error.

That is why, to me, it makes the most sense to create an internal unique identifier which we can use to validate if the object is actually new or not. **_Keep in mind the UUID is only meant for internal purposes._** 

**Just because the id is the same does not mean it isn't new data.**  Using a user provided id as the key is not a good idea (you run into this exact issue) especially when you're using that id to determine whether or not data has changed.

I created a demo that lets me edit rows.  After editing a row, our UUID remains the same but our table is still updated with edited data. That is because we are better detecting when to actually create a new key(uuid). 

<img width="288" alt="Screen Shot 2021-07-26 at 7 56 01 PM" src="https://user-images.githubusercontent.com/21092343/127078259-2385ae7b-b6f6-43f2-9c7a-7f37449e5c3b.png">

**When data from outside the table is given to us (like when data is switched in the provided example by OP) we 100% want to generate new internal UUID's for those rows, _regardless of the user provided id._** When someone is only editing a row, we don't necessarily need a new key, so we don't make one (state is changed internally, we are not being fed new data externally).

The reason we havent seen this before is because OP is providing data from **_outside_** the table which **_has changed_** but has **_retained_** the **_same id_**. Since we are using that id as the key, React says "hmm new data, same key, already exists, throw error". It's just a unique scenario, an edge case if you will.

I could be missing something, though.

**Comment by Domino987** [comment]
Sorry, had to go to bed xD
So I think we are mixing two problems a bit.
1. The key passed to the rows
We want to pass the same key to the row, if its the same data as before (even if the values of that data changed), but we want a new key if its a new object. Previously, this was always the index but now with the removal of the mutation the persistence can be tricky. This could be causing unnecessary remounts like causing state lose like edit or detail panel, even though the same data was passed after an external/parent rerender. So it does not matter if we use an internal or external id to pass as key. As to make it easier, we should just stick with one.

So far so good.
2. How  do we get a stable key for the same data
We are facing the same problem as react is: [Link](https://twitter.com/dan_abramov/status/1415279090446204929).
How to determine whether its the same object or a new one:
Creating a new id on every eternal render, if no id is provided, will cause a state loss because this guarantee a state loss all the time. Because we do not know if item at index 0 is the same object as before, we would create a new id every time. So that does not work.
If the users do not provide an id, the index is fine as long as the external component is not rerendred because we can just assume its the same object as before and no drastic changes happen.
If the data set changes or the external component updates and the use experience state loss, they must provide an id so we can track the data over multiple renders. This should work now with 
```
 const prevTableData = prevDataObject[row.id] || {};
const tableData = {
        id: row.id || index,
        ...prevTableData,
        ...row.tableData
      };
```

> When data from outside the table is given to us (like when data is switched in the provided example by OP) we 100% want to generate new internal UUID's for those rows, regardless of the user provided id

Why do you think that? Why not take the id from the user. It must be unique (since its an id) and we should just use this. If an internal Id would be enough, react could just use that as well. But since it does not, an external id is required. You example with the uuid only works, because we are merging the tableData with the help of the id. And that would be the same as using the id in the first place. But if no id provided, we see a remount on every render for every row, even if the data does not change at all and that would be a breaking case for users without an id.

> Since we are using that id as the key, React says "hmm new data, same key, already exists, throw error".

This is actually not the case. We had a bug with `prevDataObject[row.tableData && row.tableData.id]`  instead of 
`const prevTableData = prevDataObject[row.id] || {};` which caused the same id to be present multiple times resulting in the row duplication.

**Comment by edgar-arroyo-by** [comment]
> > @Domino987  Maybe just something like
> > `keyAccesor: string | () => string`
> 
> Would you mind creating a PR for this? Looks good to me to get the key. Once this is done, we need to fix the duplicates if that still happens.

Sorry sir, I got missed on the conversation.

Which is the approach that we will be following?
Is there a PR already for that approach? (I can happily work on that in case this answer is negative)


thanks in advance

**Comment by matthewoestreich** [comment]
So @Domino987 the resolution does not cause any extra renders or mounts, period. It just doesn't. You can test it if you'd like.

Even if we take the unique ID from the user, or use the id via the index, we will still run into this issue. We cannot use the user provided id as the key or else you will run into this issue because data may have changed but the key hasn't. That is exactly why we are getting this error.  

It seems like we are way overcomplicating something that is rather simple.

> This is actually not the case. We had a bug with prevDataObject[row.tableData && row.tableData.id] instead of
> const prevTableData = prevDataObject[row.id] || {}; which caused the same id to be present multiple times resulting in the row duplication.

This is not correct. I made these changes as you suggested and it only made things worse. I get the errors right away now... I don't understand - it's as simple as receiving the same key for different data. I literally have shown why this is happening, and how to prevent it.

>Creating a new id on every eternal render, if no id is provided, will cause a state loss because

If you're providing new *external* data, then the state is "lost" regardless. You are rendering with all new state. Because you have to. Because it's new state. But thats the exact issue. Since our state has technically changed but the id has not changed, we get the error.  

It's not the same issue/error as Link because I showed how we can better detect when new external data is received. 

Can you provide a sandbox that shows how those changes cause rerenders and unnecessary mounts? I have a POC showing how my changes fix this issue and address our concerns. I think it would be easier to see the issues you are talking about in a POC, otherwise, I don't see how this is causing any issues.

**Comment by Domino987** [comment]
There is no PR yet. So if you could create one that would be great.

**Comment by Domino987** [comment]
@oze4 did you try it without the id passed from the user? But I will create a sandbox

**Comment by matthewoestreich** [comment]
I made this branch which contains all the changes I made. https://github.com/material-table-core/core/tree/uuid-for-row-key

**Comment by matthewoestreich** [comment]
Yes I tested without passing the id from the user as the key.

**Comment by Domino987** [comment]
So in your branch, if I do not pass an id, every row remounts every time:
![Screencast from 27 07 2021 16_07_06](https://user-images.githubusercontent.com/17567991/127168668-cf2b6f07-4c7d-4050-9e63-af876c7b1477.gif)

**Comment by matthewoestreich** [comment]
What do you mean by pass an ID though? Where are you not passing the id?

**Comment by matthewoestreich** [comment]
And what you're showing is the exact behavior we want. Everything has to remount/render when you click that button because new data is being passed to us **from outside of the table**. This is expected behavior.

**Comment by Domino987** [comment]
From the user. So the internal Id is generated every time. I am working on more videos.

**Comment by matthewoestreich** [comment]
If you change it back to how it was originally (and still log renders) you'll see there aren't any extra renders or logs created.

**Comment by matthewoestreich** [comment]
> So the internal Id is generated every time.

It's not created every time, it's only generated when we get new data from outside the table. If you click edit on a row, change something, then save, you'll see a new UUID **does not** get generated. 

Because we didn't receive new data from outside the table. The new data came from **inside** the table (via the edit).

**Comment by edgar-arroyo-by** [comment]
@oze4 

Jusk kindly asking, 

Is you branch going to be merged on master?

I am sorry, i dunot know the development  cycle over here

**Comment by matthewoestreich** [comment]
I think once we all agree on a solution we will merge that solution.

I still want to understand where @Domino987 is coming from bc I could very well be missing something and be wrong about this.

**Comment by edgar-arroyo-by** [comment]
awesome! you both rock! 

@oze4  @Domino987

**Comment by Domino987** [comment]
Ok I think I found the problem we are talking about. 
I overlooked that you also want to persist the uuid by using the merge of tableData I proposed if an Id is provided.
If this would not be the case, the rerenders would be triggered more often, but because we can persit them with the row.id, its fine. So sorry for missing this part.
And after checking your example, the edit state is not persisted even with a provided id.
But this is not caused by the key as I thought but instead of some internal mutation of the lastEdit object with is a reference to the tableData object a row.
So yes I like your solution and the combination of using the id to track the tableData and decoupeling it from the id is fine.
Regarding the data updated but mutated, I don't think that is a problem because the row will get reredned with the new data and will display it appropriately even with the same id.
The only thing that the custom uuid fixes, is if a low number id is provided (as in the codesandbox examples at the top), which might result in key overlap, if the index is used.
So to prevent the overlap of the id with the index, i thing your UUID proposal fixes this. And because we persist it besides the external id, the behavior should be similar.

**Comment by matthewoestreich** [comment]
No need to apologize bro! That's why we have these discussions :smile: I think that you are right, because we aren't coupling with the id, that's the 'sweet spot' that doesn't make it interfere.

I should be able to add these changes today, but the question is now : should we use the uuid library to generate UUID's? I think so - I trust it more than some random function I found online.

Thoughts?

**Comment by Domino987** [comment]
That would be great if w could add that soon because the current release is a bit broken.

I think either is fine. Because there is not too much data objects anyway aka not a db, anything that generates a UUID should be fine because we have just a few obejcts.

In my projects I use uuid with 50 Mio downloads per week. I think that would be a good choice 😜

**Comment by edgar-arroyo-by** [comment]
hi there, just adding a comment over here.

Lets be aware that uuidv4 is going to be deprecated :)

**Comment by matthewoestreich** [comment]
@edgar-arroyo-by any suggestions? @Domino987 were y'all aware of the deprecation? If so, what are y'all switching to?

**Comment by Domino987** [comment]
I am not. I guess uuidv5 if it comes to it. Do you have a link to the deprecation?

**Comment by edgar-arroyo-by** [comment]
Yep, it is placed over here: https://www.npmjs.com/package/uuidv4

![image](https://user-images.githubusercontent.com/79274227/127190833-bbf377b6-7c75-4240-88ec-8fb83e0ac7eb.png)



they are deprecating in favor of https://www.npmjs.com/package/uuid

**Comment by matthewoestreich** [comment]
Ok cool so it looks like uuidv4 suggest using uuid (which is prob what @Domino987  is using).

I will use the uuid package, not uuidv4.

**Comment by Domino987** [comment]
Ah ok. Good that I am using UUID already xD

**Comment by matthewoestreich** [comment]
Going to close this out for now. You should see these changes in `v3.2.0`

**Comment by edgar-arroyo-by** [comment]
yep! thanks a lot team!
